OffchainLabs/arbitrum-token-bridge
# chore: useAllowance + better approval e2e

Closes https://github.com/OffchainLabs/arbitrum-token-bridge/issues/241

Features:
- `useAllowance` hook
- more dynamic transfer button now differentiating between approval and transfer
- better e2e tests without arbitrary wait
- also identified and fixed a bug (by using `useAllowance`) where tokens would require you to approve multiple times (after each transfer) because of `getL1TokenData` caching

Steps to test:
1. Have an unapproved token. The easiest way to get it is to start e2e locally and grab the deployed ERC-20 address (it will now log the address in the console).
2. Connect to localhost network and import the token. When you insert transfer amount, the transfer button should now say 'Approve & Deposit'. _(Note that we are broke and only have `0.000000000005` of the token - the transfer panel will show 0. Therefore you will have to work with those fractions)_
3. It should now go through the standard transfer flow, and after approving the button should say 'Move funds to L2' instead.
4. You can also test by approving a fraction of owned ERC-20, e.g. 0.00.....02, and trying to transfer 0.00.....03 (it should require you to approve more) or transferring 0.00.....01 twice - the transfer button should again say 'Approve & Deposit' when you run out of approved tokens.
